### PR TITLE
v3 migration - clusters, endpoints

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
# Cluster / Endpoints / TLS / Metadata V3 migration

This PR enables generation of V3 config for Cluster and Endpoints

# Concept of V3 migration

Most of the go-control-plane V2 imports is in `pkg/xds/envoy` and are structures are hidden behind builders.
Then, builders are used in proxy generators which don't really care about the actual type, they care that it is a resource that can be added to `ResourceSet` which will be put into the cache. You will have to now pass the API Version to the builder as a mandatory parameter so the builder can choose which version of the resource it will return.

V2 -> V3 migration will result in the copy-pasting of a massive amount of the code, but at least here it is hidden behind builders, so the copy-paste will be in one place and it will be easy to remove in the future.

Of course, it's not that obvious, V2 imports are spread around a code
image
some of the tricky parts will be
* maintaining two SnapshotCaches and servers
* Callbacks that supports V2 and V3
* proxy template modifications

but here is a good start for the actual V3 config generation

### CODE REVIEW

**Comment by nickolaev** [comment]
So far this approach is great. We'll support v2 and v3 and smoothly migrate.

File: pkg/xds/envoy/clusters/client_side_tls_configurer.go [hunk 1]
```diff
@@ -1,8 +1,10 @@
 package clusters
 
 import (
-	envoy_api "github.com/envoyproxy/go-control-plane/envoy/api/v2"
-	envoy_core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	envoy_api_v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	envoy_core_v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	envoy_cluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	envoy_core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
```
Comment (nickolaev) [inline pkg/xds/envoy/clusters/client_side_tls_configurer.go:7]: As agreed, let's always use a numbered version of the envoy imports so `envoy_core_v3` etc.

File: pkg/xds/envoy/clusters/cluster_builder.go [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
+	"errors"
 
-// ClusterConfigurer is responsible for configuring a single aspect of the entire Envoy cluster,
-// such as filter chain, transparent proxying, etc.
-type ClusterConfigurer interface {
-	// Configure configures a single aspect on a given Envoy cluster.
-	Configure(cluster *envoy_api.Cluster) error
-}
+	envoy_api_v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	envoy_api "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
```
Comment (nickolaev) [inline pkg/xds/envoy/clusters/cluster_builder.go:7]: This seems inconsistent with the below `v2` and `v3` pointing out.

Shall e just use `envoy_api_v3`? I don't mind either way, but maybe being explicit will make supporting these through thime slightly easier.

File: pkg/xds/envoy/clusters/v2/client_side_tls_configurer.go [hunk 3]
```diff
@@ -5,29 +5,22 @@ import (
 	envoy_core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
 	pstruct "github.com/golang/protobuf/ptypes/struct"
 
-	"github.com/kumahq/kuma/pkg/xds/envoy/tls"
-
 	"github.com/kumahq/kuma/pkg/core/xds"
 	"github.com/kumahq/kuma/pkg/util/proto"
-	"github.com/kumahq/kuma/pkg/xds/envoy"
+	envoy_metadata "github.com/kumahq/kuma/pkg/xds/envoy/metadata/v2"
+	tls_v2 "github.com/kumahq/kuma/pkg/xds/envoy/tls/v2"
```
Comment (nickolaev) [inline pkg/xds/envoy/clusters/v2/client_side_tls_configurer.go:11]: I would assume that under `v2` folder you don't mix versions so `tls_v2` could just be `tls`.

File: pkg/xds/envoy/clusters/v3/client_side_mtls_configurer.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	envoy_core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	pstruct "github.com/golang/protobuf/ptypes/struct"
+
+	core_xds "github.com/kumahq/kuma/pkg/core/xds"
+	"github.com/kumahq/kuma/pkg/util/proto"
+	xds_context "github.com/kumahq/kuma/pkg/xds/context"
+	"github.com/kumahq/kuma/pkg/xds/envoy"
+	envoy_metadata "github.com/kumahq/kuma/pkg/xds/envoy/metadata/v3"
+	"github.com/kumahq/kuma/pkg/xds/envoy/tls"
+	tls_v3 "github.com/kumahq/kuma/pkg/xds/envoy/tls/v3"
```
Comment (nickolaev) [inline pkg/xds/envoy/clusters/v3/client_side_mtls_configurer.go:14]: Same here, `tls` instead of `tls_v3`

File: pkg/xds/envoy/tls/v3/tls.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
+package v3
+
+import (
+	envoy_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoy_grpc_credential_v3 "github.com/envoyproxy/go-control-plane/envoy/config/grpc_credential/v3"
+	envoy_tls_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/transport_sockets/tls/v3"
+	envoy_type_matcher_v3 "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
```
Comment (nickolaev) [inline pkg/xds/envoy/tls/v3/tls.go:7]: everything here is v3, so no need to explicitly `envoy_tls_v3`

File: pkg/xds/generator/inbound_proxy_generator.go [hunk 6]
```diff
@@ -32,7 +32,7 @@ func (g InboundProxyGenerator) Generate(ctx xds_context.Context, proxy *model.Pr
 
 		// generate CDS resource
 		localClusterName := envoy_names.GetLocalClusterName(endpoint.WorkloadPort)
-		clusterBuilder := envoy_clusters.NewClusterBuilder().
+		clusterBuilder := envoy_clusters.NewClusterBuilder(envoy_common.APIV2).
```
Comment (nickolaev) [inline pkg/xds/generator/inbound_proxy_generator.go:35]: Should you put a `todo` on these so we don't accidentally miss something when we get to configure the version?

**Review by nickolaev — APPROVED** [review]
A couple of suggestions on my end. Nothing really binding, so you can merge as it is if you wish.  Great stuff!

**Comment by jakubdyszkiewicz** [comment]
I'll fix the imports in the next PR

**Pull request merged**